linkerd/io/src/lib.rs

```diff
@@ -50,86 +52,95 @@ impl PeerAddr for tokio::io::DuplexStream {
-
-        /// This method is to allow using `Async::poll_write_buf` even through a
-        /// trait object.
-        fn poll_write_buf_erased(
-            self: Pin<&mut Self>,
-            cx: &mut Context<'_>,
-            buf: &mut dyn Buf,
-        ) -> Poll<usize>;
+
+/// This trait is private, since its purpose is for creating a dynamic trait
+/// object, but doing so without care can to lead not getting vectored
+/// writes.
+///
+/// Instead, use the concrete `BoxedIo` type.
+pub trait Io: AsyncRead + AsyncWrite + PeerAddr + Send + internal::Sealed {
+    /// This method is to allow using `Async::polL_read_buf` even through a
+    /// trait object.
+    #[inline]
+    fn poll_read_buf_erased(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        mut buf: &mut dyn BufMut,
+    ) -> Poll<usize>
+    where
+        Self: Sized,
+    {
+        crate::poll_read_buf(cx, self, &mut buf)
     }
 
-    impl Io for tokio::net::TcpStream {
-        fn poll_write_buf_erased(
-            self: Pin<&mut Self>,
-            cx: &mut Context<'_>,
-            mut buf: &mut dyn Buf,
-        ) -> Poll<usize> {
-            self.poll_write_buf(cx, &mut buf)
-        }
-
-        fn poll_read_buf_erased(
-            self: Pin<&mut Self>,
-            cx: &mut Context<'_>,
-            mut buf: &mut dyn BufMut,
-        ) -> Poll<usize> {
-            self.poll_read_buf(cx, &mut buf)
-        }
+    /// This method is to allow using `Async::poll_write_buf` even through a
+    /// trait object.
+    #[inline]
+    fn poll_write_buf_erased(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        mut buf: &mut dyn Buf,
+    ) -> Poll<usize>
+    where
+        Self: Sized,
+    {
+        crate::poll_write_buf(cx, self, &mut buf)
     }
+}
+
+impl<I> Io for I where I: AsyncRead + AsyncWrite + PeerAddr + Send + internal::Sealed {}
+mod internal {
+    use super::Io;
+    pub trait Sealed {}
+    impl Sealed for tokio::net::TcpStream {}
+    impl<S: Io + Unpin> Sealed for tokio_rustls::server::TlsStream<S> {}
+    impl<S: Io + Unpin> Sealed for tokio_rustls::client::TlsStream<S> {}
+}
 
-    impl<S: Io + Unpin> Io for tokio_rustls::server::TlsStream<S> {
-        fn poll_write_buf_erased(
-            self: Pin<&mut Self>,
-            cx: &mut Context<'_>,
-            mut buf: &mut dyn Buf,
-        ) -> Poll<usize> {
-            self.poll_write_buf(cx, &mut buf)
-        }
-
-        fn poll_read_buf_erased(
-            self: Pin<&mut Self>,
-            cx: &mut Context<'_>,
-            mut buf: &mut dyn BufMut,
-        ) -> Poll<usize> {
-            self.poll_read_buf(cx, &mut buf)
-        }
+// Vendored from upstream Tokio implementation, which is currently private.
+// https://github.com/tokio-rs/tokio/blob/24ed874e81fa27a6505613051955127ba8ddfdf2/tokio-util/src/lib.rs#L72-L100
+//
+// TODO(eliza): if tokio makes this public, it would be nice to not vendor it.
+pub fn poll_read_buf<T: AsyncRead>(
```

Comment:
I'd be +1 to trying to get the upstream API exposed; but in lieu of taht we should at least document what it does

Reply:
It's public now, I'll just remove this.